examples/raytrace-parallel/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+    pub fn render(
+        self,
+        concurrency: usize,
+        pool: WorkerPool,
+        ctx: CanvasRenderingContext2d,
+    ) -> Result<RenderingScene, JsValue> {
+        let (tx, rx) = oneshot::channel();
+        let rx = rx.then(|_| Ok(JsValue::undefined()));
+
+        let data = Rc::new(RefCell::new(None::<Render>));
+
+        let pixels = (self.inner.width * self.inner.height) as usize;
+        let mut r = Render {
+            tx: Some(tx),
+            callback: None,
+            shared: Arc::new(Shared {
+                id: NEXT_ID.fetch_add(1, SeqCst),
+                need_update: AtomicBool::new(false),
+                scene: self.inner,
+                next_pixel: AtomicUsize::new(0),
+                remaining: AtomicUsize::new(concurrency),
+                rgb_data: Mutex::new(vec![0; 4 * pixels]),
+            }),
+            ctx,
+        };
+
+        let data2 = data.clone();
+        let callback = Closure::wrap(Box::new(move |msg: Event| -> Result<(), JsValue> {
+            let mut slot = data2.borrow_mut();
+            if let Some(mut data) = slot.take() {
+                match data.event(&msg) {
+                    Ok(true) => {}
+                    Ok(false) => *slot = Some(data),
+                    Err(e) => {
+                        *slot = Some(data);
+                        return Err(e)
+                    }
+                }
+            }
+            Ok(())
+        }) as Box<FnMut(_) -> Result<(), JsValue>>);
+
+        for worker in &pool.workers[..concurrency] {
+            let ptr_to_send = Arc::into_raw(r.shared.clone()) as u32;
+            let ptr_to_send = JsValue::from(ptr_to_send);
+            worker.post_message(&ptr_to_send)?;
+            worker.set_onmessage(Some(callback.as_ref().unchecked_ref()));
+            worker.set_onerror(Some(callback.as_ref().unchecked_ref()));
+        }
+
+        r.callback = Some(callback);
+        *data.borrow_mut() = Some(r);
+
+        Ok(RenderingScene {
+            inner: data,
+            promise: wasm_bindgen_futures::future_to_promise(rx),
+            pool,
+        })
+    }
+}
+
+#[wasm_bindgen]
+pub struct WorkerPool {
+    workers: Vec<Worker>,
+    callback: Closure<FnMut(Event)>,
+}
+
+#[wasm_bindgen]
+impl WorkerPool {
+    #[wasm_bindgen(constructor)]
+    pub fn new(max: u32) -> Result<WorkerPool, JsValue> {
+        let callback = Closure::wrap(Box::new(|event: Event| {
+            console_log!("unhandled event: {}", event.type_());
+        }) as Box<FnMut(Event)>);
+        let mut workers = Vec::new();
+        for _ in 0..max {
+            // TODO: what do do about `./worker.js`:
+            //
+            // * the path is only known by the bundler. How can we, as a
+            //   library, know what's going on?
+            // * How do we not fetch a script N times? It internally then
+            //   causes another script to get fetched N times...
+            let worker = Worker::new("./worker.js")?;
```

Comment:
This is an interesting issue. If you were in a JS file, the bundler would detect the import and rewrite to its own architecture.

Here it's just a random string, no module system will be able to detect it and importing it doesn't make any sense since we want the URL.

In the web it's easy to move/load the file at the root `/` but that won't work in Node.

it would make sense to me to have the bundler injects some runtime functions for plateform specific stuff, like generating a path/url. Unfortunately, type representation are language specific, how do we pass a string? 

To sum up: :man_shrugging:

Reply:
Maybe Host-bindings would help

Reply:
In this particular case, we could do this:

```rust
let workerjs = include_str!("worker.js");
late workerjs_b64 = web_sys::window().unwrap().btoa(workerjs)?;
let worker = Worker::new(&format!("data:text/javascript;base64,{}", workerjs_b64))?;
```

Reply:
Sorry @fitzgen I didn't ping you, the discussion is here https://github.com/webpack/webpack/issues/8235#issue-371444670.

I think that exposing the bundler's API is much better than inlining a big JS there.